### PR TITLE
Baro: improved atmospheric model for high altitude flight

### PR DESCRIPTION
**Pull request by tridge**
this replaces PR #17984 
the formula in this PR are from @hendjoshsr71 who re-worked my original PR to use a parametric system instead of a full table
the use of the new formula is only enabled if EKF double precision is enabled or building for SITL

### CODE REVIEW

File: libraries/AP_Baro/AP_Baro.cpp [hunk 1]
```diff
@@ -401,55 +401,6 @@ void AP_Baro::update_calibration()
...
-    float tempK = C_TO_KELVIN(get_ground_temperature()) - ISA_LAPSE_RATE * altitude;
-    const float eas2tas_squared = SSL_AIR_DENSITY / (pressure / (ISA_GAS_CONSTANT * tempK));
-    if (!is_positive(eas2tas_squared)) {
-        return 1.0f;
-    }
-    return sqrtf(eas2tas_squared);
-}
 
 // return air density / sea level density - decreases as altitude climbs
 float AP_Baro::get_air_density_ratio(void)
```
Comment (IamPete1) [inline libraries/AP_Baro/AP_Baro.cpp:406]: This uses baro EAS2TAS too, should be moved to the AHRS.

File: libraries/AP_Baro/AP_Baro_atmosphere.cpp [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+#endif // HAL_BARO_ATMOSPHERIC_EXTENDED || CONFIG_HAL_BOARD == HAL_BOARD_SITL
+
+/*
+  return geometric altitude difference in meters between current pressure and a
+  given base_pressure in Pascal.
+*/
+float AP_Baro::get_altitude_difference(float base_pressure, float pressure) const
+{
+#if HAL_BARO_ATMOSPHERIC_EXTENDED
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_atmosphere.cpp:271]: Shouldn't be in `HAL`.

`AP_BARO_1976_STANDARD_ATMOSPHERE_ENABLED`

File: libraries/AP_Baro/AP_Baro.h [hunk 3]
```diff
@@ -112,8 +134,10 @@ class AP_Baro
     // pressure in Pascal
     float get_sealevel_pressure(float pressure) const;
 
-    // get scale factor required to convert equivalent to true airspeed
-    float get_EAS2TAS(void);
+    // get scale factor required to convert equivalent to true
+    // airspeed. This should only be used to update the AHRS value
+    // once per loop. Please use AP::ahrs().get_EAS2TAS()
+    float get_EAS2TAS(void) const;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro.h:140]:
```suggestion
    float get_EAS2TAS_use_AHRS_instead(void) const;
```

.... or, perhaps, something more subtle

Comment (tridge) [reply libraries/AP_Baro/AP_Baro.h]: added _

File: libraries/AP_Baro/AP_Baro_atmosphere.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_Baro/AP_Baro_atmosphere.cpp]: done

File: libraries/AP_Baro/AP_Baro.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Baro/AP_Baro.cpp]: changed

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 4]
```diff
@@ -3570,6 +3570,16 @@ float AP_AHRS::get_EAS2TAS(void) const
     return 1.0;
 }
 
+// get air density / sea level density - decreases as altitude climbs
+float AP_AHRS::get_air_density_ratio(void) const
+{
+    const float eas2tas = get_EAS2TAS();
+    if (is_positive(eas2tas)) {
```
Comment (tridge) [inline libraries/AP_AHRS/AP_AHRS.cpp:3577]: could remove this check

Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS.cpp]: done

**Review by peterbarker — APPROVED** [review]

**Pull request merged**